pkg/nsx/services/firewall.go

```diff
@@ -457,56 +930,144 @@ func (service *SecurityPolicyService) updatePeerExpressions(obj *v1alpha1.Securi
-		"",
-		map[string]data.DataValue{
-			"expressions":   expressions,
-			"resource_type": data.NewStringValue("NestedExpression"),
-		},
-	)
-	group.Expression = append(group.Expression, expressionFrame)
+	expressions := service.buildGroupExpression(&group.Expression)
 
 	clusterExpression := service.buildExpression(
 		"Condition", "SegmentPort",
 		fmt.Sprintf("%s|%s", util.TagScopeNCPCluster, service.getCluster()),
-		"Tag", "EQUALS",
+		"Tag", "EQUALS", "EQUALS",
 	)
 	expressions.Add(clusterExpression)
 
 	if peer.PodSelector != nil {
 		service.addOperatorIfNeeded(expressions, "AND")
 		podExpression := service.buildExpression(
-			"Condition", "SegmentPort", util.TagScopeNCPPod, "Tag", "EQUALS")
+			"Condition", "SegmentPort", util.TagScopeNCPPod, "Tag", "EQUALS", "EQUALS")
 		expressions.Add(podExpression)
-		service.updatePortExpressions(peer.PodSelector.MatchLabels, expressions)
+
+		if peer.NamespaceSelector == nil {
+			mixedCriteria = false
+		} else {
+			mixedCriteria = true
+		}
+
+		memberType = "SegmentPort"
+		tagValueExpression = podExpression
+		matchLabels = peer.PodSelector.MatchLabels
+		matchExpressions = &peer.PodSelector.MatchExpressions
+		matchLabelsCount = len(matchLabels)
 	}
 	if peer.VMSelector != nil {
 		service.addOperatorIfNeeded(expressions, "AND")
 		vmExpression := service.buildExpression(
-			"Condition", "SegmentPort", util.TagScopeNCPVNETInterface, "Tag", "EQUALS")
+			"Condition", "SegmentPort", util.TagScopeNCPVNETInterface, "Tag", "EQUALS", "EQUALS")
 		expressions.Add(vmExpression)
-		service.updatePortExpressions(peer.VMSelector.MatchLabels, expressions)
+
+		if peer.NamespaceSelector == nil {
+			mixedCriteria = false
+		} else {
+			mixedCriteria = true
+		}
+
+		memberType = "SegmentPort"
+		tagValueExpression = vmExpression
+		matchLabels = peer.VMSelector.MatchLabels
+		matchExpressions = &peer.VMSelector.MatchExpressions
+		matchLabelsCount = len(matchLabels)
 	}
 	if peer.NamespaceSelector != nil {
-		service.updateSegmentSelectorExpressions(peer.NamespaceSelector.MatchLabels, expressions)
+		if !mixedCriteria {
+			tagValueExpression = nil
+			memberType = "Segment"
+			matchLabels = peer.NamespaceSelector.MatchLabels
+			matchExpressions = &peer.NamespaceSelector.MatchExpressions
+			matchLabelsCount = len(matchLabels)
+		} else { // Handle PodSelector or VMSelector mixed with NamespaceSelector
+			memberType = "Segment"
+			nsMatchLabels := peer.NamespaceSelector.MatchLabels
+			nsMatchExpressions := &peer.NamespaceSelector.MatchExpressions
+
+			// Validate expressions for POD/VM Selectors
+			mergedMatchExpressions = service.mergeSelectorMatchExpression(*matchExpressions)
+			opInValueCount, err = service.validateSelectorOpIn(*mergedMatchExpressions)
+
+			nsMergedMatchExpressions := service.mergeSelectorMatchExpression(*nsMatchExpressions)
+			nsOpInValCount, opErr := service.validateSelectorOpIn(*nsMergedMatchExpressions)
+
+			if err == nil && opErr == nil {
```

Comment:
Please change line 1011 to `if err == nil || opErr == nil` to reduce the indent for line 1012~1033 for the readibility.

Reply:
Definitely right direction to go!! thanks